README.md
[![Build Status](https://circleci.com/gh/ybrliiu/p5-Sub-WrapInType.svg)](https://circleci.com/gh/ybrliiu/p5-Sub-WrapInType) [![Coverage Status](http://codecov.io/github/ybrliiu/p5-Sub-WrapInType/coverage.svg?branch=master)](https://codecov.io/github/ybrliiu/p5-Sub-WrapInType?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/Sub-WrapInType.svg)](https://metacpan.org/release/Sub-WrapInType)
# NAME

Sub::WrapInType - Wrap the subroutine to validate the argument type and return type.

# SYNOPSIS

    use Types::Standard -types;
    use Sub::WrapInType;

    my $sum = wrap_sub [ Int, Int ], Int, sub {
      my ($x, $y) = @_;
      $x + $y;
    };
    $sum->(2, 5);  # Returns 7
    $sum->('foo'); # Throws an exception

    my $subtract = wrap_sub [ Int, Int ], Int, sub {
      my ($x, $y) = @_;
      "$x - $y";
    };
    $subtract->(5, 2); # Returns string '5 - 2', error!

# DESCRIPTION

Sub::WrapInType is wrap the subroutine to validate the argument type and return type.

# FUNCTIONS

## wrap\_sub(\\@parameter\_types, $return\_type, $subroutine)

If you pass type constraints of parameters, a return type constraint, and a subroutine to this function,
Returns the subroutine wrapped in the process of checking the arguments given in the parameter's type constraints and checking the return value with the return value's type constraint.

    my $sum = wrap_sub [Int, Int], Int, sub {
      my ($x, $y) = @_;
      $x + $y;
    };
    $sum->(2, 5);  # Returns 7
    $sum->('foo'); # Throws an exception (Can not pass string)

    my $wrong_return_value = wrap_sub [Int, Int], Int, sub {
      my ($x, $y) = @_;
      "$x + $y";
    };
    $wrong_return_value->(2, 5); # Throws an exception (The return value isn't an Integer)
    $sum->('foo');               # Throws an exception (Can not pass string)

The type constraint expects to be passed an object of Type::Tiny.

When the subroutine returns multiple return values, it is possible to specify multiple return type constraints.

    my $multi_return_values = wrap_sub [Int, Int], [Int, Int], sub {
      my ($x, $y) = @_;
      ($x, $y);
    };
    my ($x, $y) = $multi_return_values->(0, 1);

You can pass named parameters.

    my $sub = wrap_sub(
      params => [Int, Int],
      return => Int,
      code   => sub {
        my ($x, $y) = @_;
        $x + $y;
      },
    );

If the <PERL\_NDEBUG> or the <NDEBUG> environment variable is true, the subroutine will not check the argument type and return type.

If subroutine returns array or hash, Sub::WrapInType will not be able to check the type as you intended.
You should rewrite the subroutine to returns array reference or hash reference.

Sub::WrapInType does not support wantarray.

## wrap\_method(\\@parameter\_types, $return\_type, $subroutine)

This function skips the type check of the first argument:

    sub add {
      my $class = shift;
      my ($x, $y) = @_;
      $x + $y;
    }

    my $sub = wrap_method [Int, Int], Int, \&add;
    $sub->(__PACKAGE__, 1, 2); # => 3

# METHODS

## new(\\@parameter\_types, $return\_type, $subroutine, $options)

Constract a new Sub::WrapInType object.

    use Types::Standard -types;
    use Sub::WrapInType;
    my $wraped_sub = Sub::WrapInType->new([Int, Int] => Int, sub { $_[0] + $_[1] });

You can pass options.

- **check**

    Default: true

    The created subroutine check the argument type and return type.

    If you don't want to check the argument type and return type, pass false.

- **skip\_invocant**

    Default: false

    The created subroutine skips the type check of the first argument.

# LICENSE

Copyright (C) mp0liiu.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

mp0liiu <mpoliiu@cpan.org>

# SEE ALSE

[Type::Params](https://metacpan.org/pod/Type%3A%3AParams) exports the function wrap\_subs. It check only parameters type.


Changes
Revision history for Perl extension Sub-WrapInType

{{$NEXT}}

0.08 2023-04-09T00:25:58Z
    - Update Type::Tiny version

0.07 2021-03-14T02:46:50Z
    - Fix install_method to install the typed method

0.06 2021-03-12T22:39:27Z
    - Add install_sub function
    - Add install_method function

0.05 2021-03-07T20:03:29Z
    - Support skip_invocant option and add wrap_method function (kfly8++)
    - Support check option 
    - Type checking can be controlled through the PERL_NDEBUG or NDEBUG environment variables.

0.04 2020-12-12T14:15:19Z
    - Fix the version of Type::Tiny

0.03 2020-09-27T12:41:30Z

    - New format support `wrap_sub($param, ...)`

0.02 2020-09-26T17:52:53Z

    - support multiple return values

0.01_02 2020-02-24T20:11:41Z

    - fix module name

0.01_01 2020-02-24T19:50:34Z

    - original version



lib/AnonSub/Typed.pm
package AnonSub::Typed;
use 5.010001;
use strict;
use warnings;
use Carp ();
use Hash::Util ();
use Scalar::Util ();
use Type::Params ();
use Exporter qw( import );
use Class::InsideOut qw( register readonly id );

our $VERSION   = '0.01';
our @EXPORT    = qw( anon );
our @EXPORT_OK = qw( get_info );

readonly params  => my %params;
readonly returns => my %returns;
readonly code    => my %code;

sub new {
  my $class = shift;

  my ($params_types, $return_type, $code) = do {
    if (@_ == 3) {
      @_;
    }
    elsif (@_ == 6) {
      my %args = @_;
      my @valid_args = grep { defined } map { $args{$_} } qw( params isa code );
      @valid_args == 3 ? @valid_args : Carp::croak 'Wrong arguments.';
    }
    else {
      Carp::croak 'Wrong arguments.';
    }
  };

  {
    my @types = do {
      if (ref $params_types eq 'ARRAY') {
        @$params_types;
      }
      elsif (ref $params_types eq 'HASH') {
        values %$params_types;
      }
      else {
        Carp::croak 'Wrong arguments.'
      }
    };
    for my $type (@types, $return_type) {
      Carp::croak 'Wrong arguments.'
        unless Scalar::Util::blessed($type) && $type->can('check') && $type->can('get_message');
    }
  }

  my $checker = ref $params_types eq 'ARRAY'
    ? Type::Params::compile(@$params_types)
    : Type::Params::compile_named(%$params_types);

  my $typed_code = sub {
    my $return_value = $code->( $checker->(@_) );
    die $return_type->get_message($return_value) unless $return_type->check($return_value);
    $return_value;
  };

  my $self = bless $typed_code, $class;
  register($self);

  {
    my $addr = id $self;
    $params{$addr} = $params_types;
    $returns{$addr}    = $return_type;
    $code{$addr}   = $code;
  }

  $self;
}

sub anon {
  __PACKAGE__->new(@_);
}

1;

__END__

=encoding utf-8

=head1 NAME

AnonSub::Typed - Create simple typed anonymous subroutine easily.

=head1 SYNOPSIS

    use Test2::V0;
    use AnonSub::Typed;

    my $sum = anon [ Int, Int ], Int, sub {
      my ($x, $y) = @_;
      $x + $y;
    };
    is $sum->(2, 5), 7;
    done_testing;

=head1 DESCRIPTION

AnonSub::Typed is create simple typed anonymous subroutine easily.

=head1 LICENSE

Copyright (C) ybrliiu.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ybrliiu E<lt>raian@reeshome.orgE<gt>

=cut



lib/Sub/TypedAnon.pm
package Sub::TypedAnon;
use 5.010001;
use strict;
use warnings;
use Carp ();
use Hash::Util ();
use Scalar::Util ();
use Type::Params ();
use Exporter qw( import );

our $VERSION   = '0.01';
our @EXPORT    = qw( anon );
our @EXPORT_OK = qw( get_info );

sub anon {
  my ($params_types, $return_type, $code) = do {
    if (@_ == 3) {
      @_;
    }
    elsif (@_ == 6) {
      my %args = @_;
      my @valid_args = grep { defined } map { $args{$_} } qw( params isa code );
      @valid_args == 3 ? @valid_args : Carp::croak 'Wrong arguments.';
    }
    else {
      Carp::croak 'Wrong arguments.';
    }
  };

  {
    my @types = do {
      if (ref $params_types eq 'ARRAY') {
        @$params_types;
      }
      elsif (ref $params_types eq 'HASH') {
        values %$params_types;
      }
      else {
        Carp::croak 'Wrong arguments.'
      }
    };
    for my $type (@types, $return_type) {
      Carp::croak 'Wrong arguments.'
        unless Scalar::Util::blessed($type) && $type->can('check') && $type->can('get_message');
    }
  }

  my $checker = ref $params_types eq 'ARRAY'
    ? Type::Params::compile(@$params_types)
    : Type::Params::compile_named(%$params_types);

  my $typed_code = sub {
    my $return_value = $code->( $checker->(@_) );
    die $return_type->get_message($return_value) unless $return_type->check($return_value);
    $return_value;
  };

  _store_info($typed_code, $params_types, $return_type, $code);

  $typed_code;
}

{
  my %Info;

  sub _store_info {
    my ($typed_code, $params_types, $return_type, $code) = @_;
    $Info{ $typed_code . '' } = Hash::Util::lock_hashref({
      isa    => $return_type,
      params => $params_types,
      code   => $code,
    });
  }

  sub get_info {
    my $typed_code = shift;
    Carp::croak 'Wrong arguments.' unless defined $typed_code && ref $typed_code eq 'CODE';
    $Info{ $typed_code . '' };
  }
}

1;

__END__

=encoding utf-8

=head1 NAME

Sub::TypedAnon - Create simple typed anonymous subroutine easily.

=head1 SYNOPSIS

    use Test2::V0;
    use Sub::TypedAnon;

    my $sum = anon [ Int, Int ], Int, sub {
      my ($x, $y) = @_;
      $x + $y;
    };
    is $sum->(2, 5), 7;
    done_testing;

=head1 DESCRIPTION

Sub::TypedAnon is create simple typed anonymous subroutine easily.

=head1 LICENSE

Copyright (C) ybrliiu.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ybrliiu E<lt>raian@reeshome.orgE<gt>

=cut



lib/Sub/WrapInType.pm
package Sub::WrapInType;
use 5.010001;
use strict;
use warnings;
use parent 'Exporter';
use Class::InsideOut qw( register readonly id );
use Types::Standard -types;
use Type::Params qw( signature );
use Sub::Util qw( set_subname );
use namespace::autoclean;

our $VERSION = '0.07';
our @EXPORT  = qw( wrap_sub wrap_method install_sub install_method );

readonly params    => my %params;
readonly returns   => my %returns;
readonly code      => my %code;
readonly is_method => my %is_method;

my $TypeConstraint  = HasMethods[qw( check get_message )];
my $ParamsTypes     = $TypeConstraint | ArrayRef[$TypeConstraint] | Map[Str, $TypeConstraint];
my $ReturnTypes     = $TypeConstraint | ArrayRef[$TypeConstraint];
my $Options         = Dict[
  skip_invocant => Optional[Bool],
  check         => Optional[Bool],
];
my $DEFAULT_OPTIONS = +{
  skip_invocant => 0,
  check         => 1,
};

sub new {
  state $check = signature(
    method => 1,
    multi => [
      +{
        positional => [
          $ParamsTypes,
          $ReturnTypes,
          CodeRef,
          $Options, +{ default => sub { $DEFAULT_OPTIONS } }
        ],
      },
      +{
        named_to_list => 1,
        named => [
          params  => $ParamsTypes,
          isa     => $ReturnTypes,
          code    => CodeRef,
          options => $Options, +{ default => sub { $DEFAULT_OPTIONS } },
        ],
      },
    ],
  );
  my ($class, $params_types, $return_types, $code, $options) = $check->(@_);
  $options = +{ %$DEFAULT_OPTIONS, %$options };

  my $typed_code =
      $options->{check}
    ? $class->_create_typed_code($params_types, $return_types, $code, $options)
    : sub { $code->(@_) };

  my $self = bless $typed_code, $class;
  register($self);

  {
    my $addr = id $self;
    $params{$addr}    = $params_types;
    $returns{$addr}   = $return_types;
    $code{$addr}      = $code;
    $is_method{$addr} = !!$options->{skip_invocant};
  }

  $self;
}

sub _create_typed_code {
  my ($class, $params_types, $return_types, $code, $options) = @_;
  my $params_types_checker =
      ref $params_types eq 'ARRAY' ? signature(positional => $params_types)
    : ref $params_types eq 'HASH'  ? signature(named => [%$params_types], bless => 0)
    :                                signature(positional => [$params_types]);
  my $return_types_checker =
    ref $return_types eq 'ARRAY' ? signature(positional => $return_types)
    :                              signature(positional => [$return_types]);

  if ( ref $return_types eq 'ARRAY' ) {
    if ( $options->{skip_invocant} ) {
      sub {
        my @return_values = $code->( shift, $params_types_checker->(@_) );
        $return_types_checker->(@return_values);
        @return_values;
      };
    }
    else {
      sub {
        my @return_values = $code->( $params_types_checker->(@_) );
        $return_types_checker->(@return_values);
        @return_values;
      };
    }
  }
  else {
    if ( $options->{skip_invocant} ) {
      sub {
        my $return_value = $code->( shift, $params_types_checker->(@_) );
        $return_types_checker->($return_value);
        $return_value;
      };
    }
    else {
      sub {
        my $return_value = $code->( $params_types_checker->(@_) );
        $return_types_checker->($return_value);
        $return_value;
      };
    }
  }
}

sub _is_env_ndebug {
  $ENV{PERL_NDEBUG} || $ENV{NDEBUG};
}

sub wrap_sub {
  state $check = signature(
    message => << 'EOS',
USAGE: wrap_sub(\@parameter_types, $return_type, $subroutine)
    or wrap_sub(params => \@params_types, returns => $return_types, code => $subroutine)
EOS
    multi => [
      +{ positional => [ $ParamsTypes, $ReturnTypes, CodeRef ] },
      +{
        named_to_list => 1,
        named => [
          params => $ParamsTypes,
          isa    => $ReturnTypes,
          code   => CodeRef,
        ],
      },
    ],
  );
  my ($params_types, $return_types, $code) = $check->(@_);

  __PACKAGE__->new($params_types, $return_types, $code, +{ check => !_is_env_ndebug() });
}

sub wrap_method {
  state $check = signature(
    message => << 'EOS',
USAGE: wrap_method(\@parameter_types, $return_type, $subroutine)
    or wrap_method(params => \@params_types, returns => $return_types, code => $subroutine)
EOS
    multi => [
      +{ positional => [ $ParamsTypes, $ReturnTypes, CodeRef ] },
      +{
        named_to_list => 1,
        named => [
          params => $ParamsTypes,
          isa    => $ReturnTypes,
          code   => CodeRef,
        ],
      },
    ],
  );
  my ($params_types, $return_types, $code) = $check->(@_);

  my $options = +{
    skip_invocant => 1,
    check         => !_is_env_ndebug(),
  };
  __PACKAGE__->new($params_types, $return_types, $code, $options);
}

sub install_sub {
  state $check = signature(
    message => << 'EOS',
USAGE: install_sub($name, \@parameter_types, $return_type, $subroutine)
    or install_sub(name => $name, params => \@params_types, returns => $return_types, code => $subroutine)
EOS
    multi => [
      +{ positional => [ Str, $ParamsTypes, $ReturnTypes, CodeRef ] },
      +{
        named_to_list => 1,
        named => [
          name   => Str,
          params => $ParamsTypes,
          isa    => $ReturnTypes,
          code   => CodeRef,
        ],
      },
    ],
  );
  my ($name, $params_types, $return_types, $code) = $check->(@_);

  _install($name, wrap_sub($params_types, $return_types, $code), scalar caller);
}

sub install_method {
  state $check = signature(
    message => << 'EOS',
USAGE: install_method($name, \@parameter_types, $return_type, $subroutine)
    or install_method(name => $name, params => \@params_types, returns => $return_types, code => $subroutine)
EOS
    multi => [
      +{ positional => [ Str, $ParamsTypes, $ReturnTypes, CodeRef ] },
      +{
        named => [
          name   => Str,
          params => $ParamsTypes,
          isa    => $ReturnTypes,
          code   => CodeRef,
        ],
        named_to_list => 1,
      },
    ],
  );
  my ($name, $params_types, $return_types, $code) = $check->(@_);

  _install($name, wrap_method($params_types, $return_types, $code), scalar caller);
}

sub _install {
  my ($name, $code, $pkg) = @_;
  my $fullname = "${pkg}::${name}";
  {
    no strict 'refs';
    *{$fullname} = $code;
  }
  set_subname($fullname, $code);
}

1;

__END__

=encoding utf-8

=head1 NAME

Sub::WrapInType - Wrap the subroutine to validate the argument type and return type.

=head1 SYNOPSIS

  use Types::Standard -types;
  use Sub::WrapInType;

  my $sum = wrap_sub [ Int, Int ], Int, sub {
    my ($x, $y) = @_;
    $x + $y;
  };
  $sum->(2, 5);  # Returns 7
  $sum->('foo'); # Throws an exception

  my $subtract = wrap_sub [ Int, Int ], Int, sub {
    my ($x, $y) = @_;
    "$x - $y";
  };
  $subtract->(5, 2); # Returns string '5 - 2', error!

=head1 DESCRIPTION

Sub::WrapInType is wrap the subroutine to validate the argument type and return type.

=head1 FUNCTIONS

=head2 wrap_sub(\@parameter_types, $return_type, $subroutine)

If you pass type constraints of parameters, a return type constraint, and a subroutine to this function,
Returns the subroutine wrapped in the process of checking the arguments given in the parameter's type constraints and checking the return value with the return value's type constraint.

  my $sum = wrap_sub [Int, Int], Int, sub {
    my ($x, $y) = @_;
    $x + $y;
  };
  $sum->(2, 5);  # Returns 7
  $sum->('foo'); # Throws an exception (Can not pass string)

  my $wrong_return_value = wrap_sub [Int, Int], Int, sub {
    my ($x, $y) = @_;
    "$x + $y";
  };
  $wrong_return_value->(2, 5); # Throws an exception (The return value isn't an Integer)
  $sum->('foo');               # Throws an exception (Can not pass string)

The type constraint expects to be passed an object of Type::Tiny.

When the subroutine returns multiple return values, it is possible to specify multiple return type constraints.

  my $multi_return_values = wrap_sub [Int, Int], [Int, Int], sub {
    my ($x, $y) = @_;
    ($x, $y);
  };
  my ($x, $y) = $multi_return_values->(0, 1);

You can pass named parameters.

  my $sub = wrap_sub(
    params => [Int, Int],
    return => Int,
    code   => sub {
      my ($x, $y) = @_;
      $x + $y;
    },
  );

If the C<PERL_NDEBUG> or the C<NDEBUG> environment variable is true, the subroutine will not check the argument type and return type.

If subroutine returns array or hash, Sub::WrapInType will not be able to check the type as you intended.
You should rewrite the subroutine to returns array reference or hash reference.

Sub::WrapInType does not support wantarray.

=head2 wrap_method(\@parameter_types, $return_type, $subroutine)

This function skips the type check of the first argument:

  sub add {
    my $class = shift;
    my ($x, $y) = @_;
    $x + $y;
  }

  my $sub = wrap_method [Int, Int], Int, \&add;
  $sub->(__PACKAGE__, 1, 2); # => 3

=head2 install_sub($name, \@parameter_types, $return_type, $subroutine)

Install the wrapped subroutine into the current package.

  use Sub::WrapInType qw( install_sub );

  install_sub sum => [ Int, Int ] => Int, sub {
    my ($x, $y) = @_;
    $x + $y;
  };
  sum(2, 5);  # Returns 7
  sum('foo'); # Throws an exception

=head2 install_method($name, \@parameter_types, $return_type, $subroutine)

Install the wrapped method into the current package.

=head1 METHODS

=head2 new(\@parameter_types, $return_type, $subroutine, $options)

Constract a new Sub::WrapInType object.

  use Types::Standard -types;
  use Sub::WrapInType;
  my $wraped_sub = Sub::WrapInType->new([Int, Int] => Int, sub { $_[0] + $_[1] });

=head3 options

=over 2

=item *

B<< check >>

Default: true

The created subroutine check the argument type and return type.

If you don't want to check the argument type and return type, pass false.

=item *

B<< skip_invocant >>

Default: false

The created subroutine skips the type check of the first argument.

=back

=head1 LICENSE

Copyright (C) mp0liiu.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

mp0liiu E<lt>mpoliiu@cpan.orgE<gt>

=head1 SEE ALSE

L<Type::Params> exports the function wrap_subs. It check only parameters type.

=cut

